[useRenderer] Add public hook (#1418)
mui/base-ui: docs/src/app/(public)/(content)/react/utils/use-render/page.mdx

# useRender

<Subtitle>Hook for enabling a render prop in custom components.</Subtitle>

<Meta name="description" content="Hook for enabling a render prop in custom components." />

The `useRender` hook enables custom components to provide a `render` prop to override the default rendered element.

## API reference

### Input parameters

<PropsReferenceTable
  data={{
    render: {
      type: 'RenderProp<State>',
      description:
        'The React element or a function that returns one to override the default element.',
    },
    props: {
      type: 'Record<string, unknown>',
      description:
        'Props to be spread on the rendered element.\n\nThey are merged with the internal props\n of the component, so that event handlers are merged,\nclass names and styles are joined, and other external props overwrite the internal ones.',
    },

>     state: {

Reviewer: Should be `refs`*? Wonder if it's confusing that `params.props.ref` and params.refs` are both a thing?

Author: I wanted to make it simple for both scenarios, e.g. ref is already in the props and people don't have additional regs, they don't need to specify anything, if they have additional refs, they can use the refs array.

Reviewer: ```suggestion
    refs: {
```
(`state` is duplicated below)

      description: 'The refs to apply to the rendered element.',
    },
    state: {
      type: 'State',
      description:
        'The state of the component. It will be used as a parameter for the render callback.',
    },
  }}
/>

### Return value

The hook returns a `renderElement` function that returns the element to be rendered when called.

## Examples

A `render` prop for a custom Text component lets consumers use the `render` prop to replace the default rendered `p` element with a different element tag.

<Demo path="./demos/render" />

The callback version of the `render` prop also enables accessing the `state` of a component.

<Demo path="./demos/render-callback" />

### Merging props

The `mergeProps` function merges two or more sets of React props together. It safely merges three types of props:

1. Event handlers, so that all are invoked
2. `className` strings
3. `style` properties

The `ref` prop is not handled by `mergeProps`, but refs are handled when the ref is contained within the `props` object passed to `useRender`.

`mergeProps` merges objects from left to right, so that subsequent objects' propertes in the arguments overwrite previous ones' properties.

> Merging props is useful in custom components, as well as the callback version of the `render` prop for any Base UI component.

Reviewer: This should be updated based on the new `refs` option param. I also made a typo in `subsequent objects' properties`

Reviewer: I think somehow mentioning that React 19 doesn't require `forwardRef` anymore, and that the default forwarded ref simply goes inside `props` could be useful?

````suggestion
`mergeProps` merges objects from left to right, so that subsequent objects' properties in the arguments overwrite previous ones. Merging props is useful when creating custom components, as well as inside the callback version of the `render` prop for any Base UI component. 

The `ref` prop is merged separately using the `refs` option in `useRender`, which holds an array of refs.

```js
// React 19
function Text({ render = <p />, ...props }: TextProps) {
  const internalRef = React.useRef<HTMLElement | null>(null);

  const { renderElement } = useRender({
    render,
    props,
    refs: [internalRef],
  });
  
  return renderElement();
}

// React 17 and 18
const Text = React.forwardRef(function Text(
  { render = <p />, ...props }: TextProps, 
  forwardedRef: React.ForwardedRef<HTMLElement>
) {
  const internalRef = React.useRef<HTMLElement | null>(null);

  const { renderElement } = useRender({
    render,
    props,
    refs: [forwardedRef, internalRef],
  });
  
  return renderElement();
});
```
````

- [data attributes](https://base-ui.com/react/handbook/styling#data-attributes) that map to the component's state which can be used for styling.
- A [callback](/react/handbook/styling#css-classes) on the `className` prop that enables passing CSS classes depending on the component's state.

## API reference

### Input parameters

<PropsReferenceTable
  data={{
    render: {
      type: 'RenderProp<State>',
      description:
        "Allows you to replace the component's HTML element\nwith a different tag, or compose it with another component.\n\nAccepts a \`ReactElement\` or a function that returns the element to render.",
    },
    props: {
      type: 'Record<string, unknown>',
      description: 'Props to be spread on the rendered element.',
    },
    className: {
      type: 'string | ((state: State) => string)',
      description:
        "CSS class applied to the element, or a function that returns a class based on the component's state.",
    },
    state: {
      type: 'State',
      description:
        'The state of the component. It will be used as a parameter for the render and className callbacks.',
    },
    stateAttributesMap: {

>       type: 'StateAttributes<State>',

Reviewer 2: AFAIK this type is not available for users so not sure if having `StateAttributes` here is helpful.

      description: 'The refs to apply to the rendered element.',
    },
    state: {
      type: 'State',
      description:
        'The state of the component. It will be used as a parameter for the render callback.',
    },
  }}
/>

### Return value

The hook returns a `renderElement` function that returns the element to be rendered when called.

## Examples

A `render` prop for a custom Text component lets consumers use the `render` prop to replace the default rendered `p` element with a different element tag.

<Demo path="./demos/render" />

The callback version of the `render` prop also enables accessing the `state` of a component.

<Demo path="./demos/render-callback" />

### Merging props

The `mergeProps` function merges two or more sets of React props together. It safely merges three types of props:

1. Event handlers, so that all are invoked
2. `className` strings
3. `style` properties

`mergeProps` merges objects from left to right, so that subsequent objects' properties in the arguments overwrite previous ones. Merging props is useful when creating custom components, as well as inside the callback version of the `render` prop for any Base UI component.

> The `ref` prop is merged separately using the `refs` option in `useRender`, which holds an array of refs.
> 
> ```tsx
> // React 19
> function Text({ render = <p />, ...props }: TextProps) {
>   const internalRef = React.useRef<HTMLElement | null>(null);
> 
>   const { renderElement } = useRender({
>     render,
>     props,
>     refs: [internalRef],
>   });
> 
>   return renderElement();
> }
> 
> // React 17 and 18
> const Text = React.forwardRef(function Text(
>   { render = <p />, ...props }: TextProps,
>   forwardedRef: React.ForwardedRef<HTMLElement>,
> ) {
>   const internalRef = React.useRef<HTMLElement | null>(null);
> 
>   const { renderElement } = useRender({
>     render,
>     props,
>     refs: [forwardedRef, internalRef],
>   });
> 
>   return renderElement();
> });
> ```

Reviewer: ````suggestion
#### Merging refs

When building custom components, you often need to control a ref internally while still letting external consumers pass their own ref. Ref merging lets both parties have access to the underlying DOM element—the `refs` option in `useRender` enables this, which holds an array of refs to be merged together.

In the React 19, `React.forwardRef()` is not needed when building primitive components, as the external ref prop is already contained inside `props`. Your internal ref can be passed to `refs` to be merged with `props.ref`:

```tsx title="React 19"
function Text({ render = <p />, ...props }: TextProps) {
  const internalRef = React.useRef<HTMLElement | null>(null);

  const { renderElement } = useRender({
    render,
    props,
    refs: [internalRef],
  });

  return renderElement();
}
```

In older versions of React, you need to use `React.forwardRef()` and add the forwarded ref to the `refs` array along with your own internal ref:

```tsx title="React 18 and 17"
const Text = React.forwardRef(function Text(
  { render = <p />, ...props }: TextProps,
  forwardedRef: React.ForwardedRef<HTMLElement>,
) {
  const internalRef = React.useRef<HTMLElement | null>(null);

  const { renderElement } = useRender({
    render,
    props,
    refs: [forwardedRef, internalRef],
  });

  return renderElement();
});
```
````

# useRender

<Subtitle>Utility for creating custom components that support a render prop.</Subtitle>
<Meta name="description" content="Utility for adding Base UI-like features to custom components." />

The `useRender` hook allows you to support the `render` prop to override the default rendered element (similar to `asChild` from Radix) in your custom components.

## API reference

### Input parameters

<PropsReferenceTable
  data={{
    render: {
      type: 'RenderProp<State>',
      description:

>         "Allows you to replace the component's HTML element\nwith a different tag, or compose it with another component.\n\nAccepts a \`ReactElement\` or a function that returns the element to render.",

Reviewer 2: ```suggestion
        "The React element or a function that returns one to override the default element.",
```

# useRender

<Subtitle>Utility for supporting a render prop in custom components.</Subtitle>
<Meta name="description" content="Utility for supporting a render prop in custom components." />

The `useRender` hook allows you to support the `render` prop to override the default rendered element (similar to `asChild` from Radix) in your custom components.

## API reference

### Input parameters

<PropsReferenceTable
  data={{
    render: {
      type: 'RenderProp<State>',
      description:
        'The React element or a function that returns one to override the default element.',
    },
    props: {
      type: 'Record<string, unknown>',
      description:
        'Props to be spread on the rendered element.\n\nThey are merged with the internal props\n of the component, so that event handlers are merged,\nclass names and styles are joined, and other external props overwrite the internal ones.',
    },

>     state: {

Reviewer 2: Should we mention that this parameter is optional?

Author: I don't think we show optional/required parameter anywhere on the docs yet. We could think of how we want to do this separately. I think it would be better if we just mention the required params/props on the docs instead of the other way around. The type should be enough to guide people.

      type: 'State',
      description:
        'The state of the component. It will be used as a parameter for the render callback.',
    },
  }}
/>

### Return value

The hook returns a function that when called returns the element that should be rendered.

## Examples

This is an example of a Text component supporting the `render` prop.

<Demo path="./demos/render" />

> Developers can also use the callback version of the `render` prop.

Reviewer 2: Should we mention `state`?

```suggestion
Developers can also use the callback version of the `render` prop to access the internal `state` of a component.
```

## API reference

### Input parameters

<PropsReferenceTable
  data={{
    className: {
      type: 'string | ((state: State) => string)',
      description:
        "CSS class applied to the element, or a function that returns a class based on the component's state.",
    },
    render: {
      type: 'RenderProp<State>',
      description:
        "Allows you to replace the component's HTML element\nwith a different tag, or compose it with another component.\n\nAccepts a \`ReactElement\` or a function that returns the element to render.",
    },
    state: {
      type: 'State',
      description:
        'The state of the component. It will be used as a parameter for the render and className callbacks.',
    },
    props: {
      type: 'Record<string, unknown>',
      description: 'Props to be spread on the rendered element.',
    },
    stateAttributesMap: {
      type: 'StateDataAttributes<State>',
      description: 'A mapping of state to style hooks.',
    },
  }}
/>

### Return value

The hook returns a function that when called returns the element that should be rendered.

## Usage

### Supporting the `render` prop

> This is an example of a Text component that provides the support for the render prop.
> 
> <Demo path="./demos/render" />

Reviewer 2: I think this example makes it look more complex than it really is, due to imports, types, styles, etc. I wonder if it'd be better to just have a code snippet and not a live example for this. Something like:

````suggestion
This is an example on how to create your own `render` prop:

```jsx
import * as React from 'react';
import { useRender } from '@base-ui-components/react/use-render';

function Text({ render, ...otherProps }) {
  const { render, ...otherProps } = props;

  const { renderElement } = useRender({
    render: render ?? <p />,
    props: otherProps,
  });

  return renderElement();
}
```

It'd be used as follows:

```jsx
<Text>I'm a paragraph</Text>
<Text render={<span />}>I'm a span</Text>
```
````

The downside is not being able to open it on CodeSandbox, but not sure if it's that important.

Reviewer 3: It could be an (another) argument for having the ability to select code to display in a demo preview. This way we could have the best of both - a concise code snippet that's runnable and can be opened in CSB. I could take a look at this next if you agree this could be handy.

Reviewer 2: > It could be an (another) argument for having the ability to select code to display in a demo preview

This would be perfect.

> code snippets are not verified in any way (not executed, linted, typechecked, etc.), so they could be invalid (or inadvertently become invalid in the future after changes to the API)

Yeah, it's a trade-off. Sometimes live demos are an overkill and makes things harder to convey. I'll leave the live demo for now.

> ### Generating data attributes

Reviewer 2: I think we could come up with a better demo. Clicking a text is not very common and it's hard to discover. We could also simplify it and only leave one "state" (color or size, but not both).

We could also use this demo only for data attributes and leave the className callback as paragraph + code snippet in a new section. This could also improve the discoverability of the className callback docs.

Author: Yeah, the thing I was trying to add here is that people can use internal state as well, I couldn't come up with an idea for internal state for Text component :) I am ok with simplifying it.

Reviewer 2: I changed the demo to be a counter. Also added a separate demo for the className callback.

> In the following demo, the Text component provides more Base UI features, like adding data attributes and `className` callback where developers can have access to the internal state of the component.

Reviewer 2: ```suggestion
`useRender` can also be used to automatically add data attributes based on the component's internal state and support a callback in the `className` prop where developers can access the component's internal state.
```

It's weird that the paragraph talks about the className callback but the it's not present in the demo.

Author: Maybe we can extend the demo on the usage to show className callback where a class is added fort he large variant. What do you think?

Reviewer 2: The className callback is now a separate demo.

<Demo path="./demos/data-attributes" />

Sometimes, you may want to customize how the data attributes are generated. You can do this by providing the `stateAttributesMap` option.

```jsx
useRender({
  render: render ?? <p />,
  state,
  stateAttributesMap: {
    size: (value) => ({ [`data-size-${value}`]: '' }),
    color: (value) => ({ [`data-color-${value}`]: '' }),
  },
  props,
});
```

> Then, developers can target the data attributes like this:

Reviewer 2: Seeing `data-size-small` might confuse users into thinking the generated data attributes have this shape by default (instead of `data-size='small'`). I'm not sure if this paragraph and the code block below are necessary.

Author: Maybe we need a better description of what it is about.

Reviewer 2: I updated this section and removed the CSS code block.

      description:
        'The React element or a function that returns one to override the default element.',
    },
    props: {
      type: 'Record<string, unknown>',
      description:
        'Props to be spread on the rendered element.\n\nThey are merged with the internal props\n of the component, so that event handlers are merged,\nclass names and styles are joined, and other external props overwrite the internal ones.',
    },
    state: {
      type: 'State',
      description:
        'The state of the component. It will be used as a parameter for the render callback.',
    },
  }}
/>

### Return value

The hook returns a function that when called returns the element that should be rendered.

## Examples

This is an example of a Text component supporting the `render` prop.

<Demo path="./demos/render" />

Developers can also use the callback version of the `render` prop to access the internal `state` of a component.

<Demo path="./demos/render-callback" />

### Merging the props

You can use the `mergeProp` utility to safely merge the props.
It takes care of concatenating the class names, as well as merging the style prop and the event handlers.

> You can use it both in the `useRender` hook and the callback version of the render prop, e.g.

Author: Makes sense 👌

# useRender

> 

Reviewer: My take on the docs:

````mdx
# useRender

<Subtitle>Hook for enabling a `render` prop in custom components.</Subtitle>

<Meta name="description" content="Utility for supporting a render prop in custom components." />

The `useRender` hook enables custom components to provide a `render` prop to override the default rendered element.

## API reference

### Input parameters

<PropsReferenceTable
data={{
    render: {
      type: 'RenderProp<State>',
      description:
        'The React element or a function that returns one to override the default element.',
    },
    props: {
      type: 'Record<string, unknown>',
      description:
        'Props to be spread on the rendered element.\n\nThey are merged with the internal props\n of the component, so that event handlers are merged,\nclass names and styles are joined, and other external props overwrite the internal ones.',
    },
    state: {
      type: 'State',
      description:
        'The state of the component. It will be used as a parameter for the render callback.',
    },
  }}
/>

### Return value

The hook returns a `renderElement` function that returns the element to be rendered when called.

## Examples

A `render` prop for a custom Text component lets consumers use the `render` prop to replace the default rendered `p` element with a different element tag.

<Demo path="./demos/render" />

The callback version of the `render` prop also enables accessing the `state` of a component.

<Demo path="./demos/render-callback" />

### Merging props

The `mergeProps` function merges two or more sets of React props together. It safely merges three types of props:

1. Event handlers, so that all are invoked
2. `className` strings
3. `style` properties

The `ref` prop and other mergeable properties like `aria-describedby` are not handled.

It merges objects from left to right, so that subsequent objects' propertes in the arguments overwrite previous ones' properties.

Merging props is useful in custom components, as well as the callback version of the `render` prop for any Base UI component.

```jsx {7-7} title="Using mergeProps in the render callback"
import { mergeProps } from "@base-ui-components/react/merge-props";
import styles from "./index.module.css";

function Button() {
  return (
    <Component
      render={(props, state) => (
        <button {...mergeProps(props, { className: styles.Button })} />
      )}
    />
  );
}
```

### Migrating from Radix

Radix uses the `asChild` prop, while Base UI uses the `render` prop. Learn more about how composition works in Base UI in the [composition guide](/react/handbook/composition).

In Radix, the `Slot` component lets you implement an `asChild` prop.

```jsx title="Custom asChild prop with Radix"
import { Slot } from "radix-ui";

function Button({ asChild, ...props }) {
  const Comp = asChild ? Slot.Root : "button";
  return <Comp {...props} />;
}

// Usage
<Button asChild>
  <a href="/contact">Contact</a>
</Button>;
```

In Base UI `useRender` lets you implement a `render` prop. The example below is the equivalent implementation to the Radix example above.

```jsx title="Custom render prop with Base UI"
import { useRender } from "@base-ui-components/react/use-render";

function Button({ render = <button />, ...props }) {
  const { renderElement } = useRender({ render, props });
  return renderElement();
}

// Usage
<Button render={<a href="/contact">Contact</a>} />;
```
````

It merges objects from left to right. Subsequent sources overwrite property assignments of previous sources.

You can use it both in the `useRender` hook and the callback version of the render prop.

```jsx {7-7} title="Using mergeProps in the render callback"
import { mergeProps } from '@base-ui-components/react/merge-props';
import styles from './index.module.css';

function Button() {
  return (
    <Component
      render={(props, state) => <button {...mergeProps(props, { className: styles.Button })} />}
    />
  );
}
```

### Migrating from Radix

Radix and Base UI take different approaches to customizing the rendered element within a component. Radix uses the `asChild` prop, while Base UI uses the `render` prop. Learn more about how composition works in Base UI in the [composition guide](/react/handbook/composition).

In Radix, you'd use the `Slot` component to implement your own `asChild` prop.

```jsx title="Custom asChild prop with Radix"
import { Slot } from 'radix-ui';

function Button({ asChild, ...props }) {
  const Comp = asChild ? Slot.Root : 'button';
  return <Comp {...props} />;
}

// Usage
<Button asChild>
  <a href="/contact">Contact</a>
</Button>;
```

> In Base UI, you use `useRender` to implement your own `render` prop. The example below is the equivalent implementation to the Radix example above.

Reviewer 2: re "The example below is the equivalent implementation to the Radix example above", do you think this example should use the `mergeProps` util to be fully equivalent to the Radix example?

Author: Hmm, the example is to simple to show the usage of mergeProps as we are not defining any props in the component itself. Maybe the other examples should go in more details and we can keep this one simple, what do you think?

Reviewer 2: 👍
